File: examples/cpp/source/optimization_solvers/lbfgs_cr_entr_loss_dense_batch.cpp
/* file: lbfgs_cr_entr_loss_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014-2019 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    C++ example of the limited memory Broyden-Fletcher-Goldfarb-Shanno
!    algorithm with cross entropy loss function
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-LBFGS_CR_ENTR_LOSS-BATCH"></a>
 * \example lbfgs_cr_entr_loss_dense_batch.cpp
 */

#include "daal.h"
#include "service.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;
using namespace daal::data_management;

const string datasetFileName = "../data/batch/logreg_train.csv";
const size_t nFeatures = 6;  /* Number of features in training and testing data sets */
const size_t nClasses = 5;  /* Number of classes */
const size_t nIterations = 1000;
const float  stepLength  = 1.0e-4f;

int main(int argc, char *argv[])
{
    checkArguments(argc, argv, 1, &datasetFileName);

    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> dataSource(datasetFileName,
                                                 DataSource::notAllocateNumericTable,
                                                 DataSource::doDictionaryFromContext);

    /* Create Numeric Tables for input data and dependent variables */
    NumericTablePtr data(new HomogenNumericTable<>(nFeatures, 0, NumericTable::doNotAllocate));
    NumericTablePtr dependentVariables(new HomogenNumericTable<>(1, 0, NumericTable::doNotAllocate));
    NumericTablePtr mergedData(new MergedNumericTable(data, dependentVariables));

    /* Retrieve the data from input file */
    dataSource.loadDataBlock(mergedData.get());

    services::SharedPtr<optimization_solver::cross_entropy_loss::Batch<> > func = optimization_solver::cross_entropy_loss::Batch<>::create(nClasses, data->getNumberOfRows());
    func->input.set(optimization_solver::cross_entropy_loss::data, data);
    func->input.set(optimization_solver::cross_entropy_loss::dependentVariables, dependentVariables);

    /* Create objects to compute LBFGS result using the default method */
    optimization_solver::lbfgs::Batch<> algorithm(func);
    algorithm.parameter.nIterations = nIterations;
    algorithm.parameter.stepLengthSequence =
        NumericTablePtr(new HomogenNumericTable<>(1, 1, NumericTableIface::doAllocate, stepLength));

    const size_t nParameters = nClasses * (nFeatures + 1);
    DAAL_DATA_TYPE initialPoint[nParameters];
    for(size_t i = 0; i < nParameters; ++i)
        initialPoint[i] = 0.001f;

    /* Set input objects for LBFGS algorithm */
    algorithm.input.set(optimization_solver::iterative_solver::inputArgument, HomogenNumericTable<>::create(initialPoint, 1, nParameters));

    /* Compute LBFGS result */
    algorithm.compute();

    DAAL_DATA_TYPE expectedPoint[nParameters] = { -2.277f, 2.836f, 14.985f, 0.511f, 7.510f, -2.831f, -5.814f, -0.033f, 13.227f, -24.447f, 3.730f,
        10.394f, -10.461f, -0.766f, 0.077f, 1.558f, -1.133f, 2.884f, -3.825f, 7.699f, 2.421f, -0.135f, -6.996f, 1.785f, -2.294f, -9.819f, 1.692f,
        -0.725f, 0.069f, -8.41f, 1.458f, -3.306f, -4.719f, 5.507f, -1.642f };

    NumericTablePtr expectedCoefficients = HomogenNumericTable<>::create(expectedPoint, 1, nParameters);

    /* Print computed LBFGS results */
    printNumericTable(expectedCoefficients,
                      "Expected coefficients:");
    printNumericTable(algorithm.getResult()->get(optimization_solver::iterative_solver::minimum),
                      "Resulting coefficients:");
    printNumericTable(algorithm.getResult()->get(optimization_solver::iterative_solver::nIterations),
                      "Number of iterations performed:");
    return 0;
}

File: examples/python/source/optimization_solvers/lbfgs_cr_entr_loss_dense_batch.py
# file: lbfgs_cr_entr_loss_dense_batch.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#===============================================================================

# !  Content:
# !    Python example of the limited memory Broyden-Fletcher-Goldfarb-Shanno
# !    algorithm with cross entropy loss function
# !*****************************************************************************

#
## <a name="DAAL-EXAMPLE-PY-LBFGS_CR_ENTR_LOSS_BATCH"></a>
##  \example lbfgs_cr_entr_loss_dense_batch.py
#

import os
import sys

import numpy as np

import daal.algorithms.optimization_solver as optimization_solver
import daal.algorithms.optimization_solver.cross_entropy_loss
import daal.algorithms.optimization_solver.lbfgs
import daal.algorithms.optimization_solver.iterative_solver

from daal.data_management import (
    DataSourceIface, FileDataSource, HomogenNumericTable, MergedNumericTable, NumericTableIface
)

utils_folder = os.path.realpath(os.path.abspath(os.path.dirname(os.path.dirname(__file__))))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable

datasetFileName = os.path.join('..', 'data', 'batch', 'logreg_train.csv')

nFeatures = 6
nClasses = 5
nIterations = 1000
stepLength = 1.0e-4

if __name__ == "__main__":

    # Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file
    dataSource = FileDataSource(datasetFileName,
                                DataSourceIface.notAllocateNumericTable,
                                DataSourceIface.doDictionaryFromContext)

    # Create Numeric Tables for input data and dependent variables
    data = HomogenNumericTable(nFeatures, 0, NumericTableIface.doNotAllocate)
    dependentVariables = HomogenNumericTable(1, 0, NumericTableIface.doNotAllocate)
    mergedData = MergedNumericTable(data, dependentVariables)

    # Retrieve the data from input file
    dataSource.loadDataBlock(mergedData)

    func = optimization_solver.cross_entropy_loss.Batch(nClasses, data.getNumberOfRows())
    func.input.set(optimization_solver.cross_entropy_loss.data, data)
    func.input.set(optimization_solver.cross_entropy_loss.dependentVariables, dependentVariables)

    # Create objects to compute LBFGS result using the default method
    algorithm = optimization_solver.lbfgs.Batch(func)
    algorithm.parameter.nIterations = nIterations
    algorithm.parameter.stepLengthSequence = HomogenNumericTable(1, 1, NumericTableIface.doAllocate, stepLength)

    # Set input objects for LBFGS algorithm
    nParameters = nClasses * (nFeatures + 1)
    initialPoint = np.full((nParameters, 1), 0.001, dtype=np.float64)
    algorithm.input.setInput(optimization_solver.iterative_solver.inputArgument, HomogenNumericTable(initialPoint))

    # Compute LBFGS result
    # Result class from daal.algorithms.optimization_solver.iterative_solver
    res = algorithm.compute()

    expectedPoint = np.array([[-2.277], [2.836], [14.985], [0.511], [7.510], [-2.831], [-5.814], [-0.033], [13.227], [-24.447], [3.730],
        [10.394], [-10.461], [-0.766], [0.077], [1.558], [-1.133], [2.884], [-3.825], [7.699], [2.421], [-0.135], [-6.996], [1.785], [-2.294], [-9.819], [1.692],
        [-0.725], [0.069], [-8.41], [1.458], [-3.306], [-4.719], [5.507], [-1.642]], dtype=np.float64)
    expectedCoefficients = HomogenNumericTable(expectedPoint)

    # Print computed LBFGS results
    printNumericTable(expectedCoefficients, "Expected coefficients:")
    printNumericTable(res.getResult(optimization_solver.iterative_solver.minimum), "Resulting coefficients:")
    printNumericTable(res.getResult(optimization_solver.iterative_solver.nIterations), "Number of iterations performed:")
